### ISSUE TITLE
Language Filter Not Working

### ISSUE THREAD

**Issue by lynkono** [tootsuite/mastodon#13891]
### Expected behaviour
The languages that I selected will be shown to me

### Actual behaviour
Languages that I did not select are still shown to me.

### Steps to reproduce the problem
Go to Settings > Preferences > Other > Language Filter and select the language you want toots to be shown. Go to a local timeline and see if the languages you selected are only there. 

###Specifications
Google Chrome on iOS
![0CC2CB98-A657-4DC9-B2DA-964650E8D0C4](https://user-images.githubusercontent.com/66178425/83335600-fac34600-a27b-11ea-8046-f75cc48b1745.jpeg)

**Comment by Sasha-Sorokin** [comment]
Could not reproduce at [mastodon.social](https://mastodon.social/). Many languages are actually incorrectly detected as English (either the post is too short or letter set is too similar) and not much users set language that they post in in Settings, this can be an issue, but the filter itself is working.

Related issues: #11118.

**Comment by lynkono** [comment]
That's strange. I just checked my local timeline and found a post in Thai, which uses a different alphabet than English. How did it detect it as English and if not, then why would the person put their account under English if that's the case?

**Comment by Sasha-Sorokin** [comment]
Can you please link the post that was incorrectly detected?

**Comment by lynkono** [comment]
https://mas.to/@mizuboshi340/104276455190708313
This was in my local timeline

**Comment by Sasha-Sorokin** [comment]
Thank you!

The problem with this post is clear: it does not meet the word count criteria (neither [it has 4 of the words separated by spaces](https://github.com/tootsuite/mastodon/blob/v3.1.4/app/lib/language_detector.rb#L39-L41), nor it matches [“reliable characters regular expression”](https://github.com/tootsuite/mastodon/blob/v3.1.4/app/lib/language_detector.rb#L43-L51)), therefore language detection being skipped in favor of account's default posting locale, which is set to default value — English.

I actually believe this is an unintended bug, giving that “The Thai language does not use spaces between the words in a sentence” ([source](http://www.thai-language.com/ref/breaking-words)). The problem can be solved by adding `\p{Thai}` to [`RELIABLE_CHARACTERS_RE`](https://github.com/tootsuite/mastodon/blob/v3.1.4/app/lib/language_detector.rb#L7).

Do you have any more examples besides Thai one?

**Comment by lynkono** [comment]
I remember seeing a lot of Spanish posts a few days ago, but I don't see any examples now. They also share the similar alphabet as English, so I don't know how that can be filtered out.

**Comment by Sasha-Sorokin** [comment]
Hi! I've got some free time and can fix issue related to Thai language. Did you find any other languages not detected correctly?

**Comment by pauby** [comment]
I have also set English as the only language to display in the filter. I see these posts just now:

* https://m.cmx.im/@anticlockwise/108227641272791118
* https://mastodon.social/@mrcsbmr/108227641589153693
* https://social.fedcast.ch/objects/967bd2c2-7e2f-4892-ac3b-e660b5f1e4df
* https://xn--y9azesw6bu.xn--y9a3aq/content/e669fabf-bce6-4f04-895a-7100dd9c8615/
* https://xn--y9azesw6bu.xn--y9a3aq/content/e246f2ba-3008-4030-9a42-257e7856c5f1/
* https://mastodon.social/@undivaga/108227636773516175

There are many more but I thought I'd stop at 6. I suspect some of these fall foul of the no spaces between words that you mentioned above.

**Comment by Brawaru** [comment]
As of #17478 language detection is no more, your posting language is defined by the interface language. I might assume these people use English language, so that's why you see these posts while using language filter.

**Comment by pauby** [comment]
If that's the case, shouldn't this issue be closed? Appreciate you may only be joining the dots on this now.

Two questions, if I can:
 
* Is there anywhere that documents how this feature is _supposed_ to work? I couldn't find anything.
* Is there any work going into making this better as 'we default to interface language' is a technical solution to 'the thing we had wasn't working' but not a solution to the problem of timelines being full of 'unreadable' messages. For me it's about 1 in 3 messages on average being unreadable. As a user, that's not sustainable for me to keep muting them.

Thanks.

**Comment by Brawaru** [comment]
Well, it filters posts by language, this is all it does and it kinda works, it's just that the post languages are not correct. As federated feed appears to be deprecated and probably going to be removed in the future, I guess this feature is slowly losing its purpose and probably gonna be gone too. However, discovery features like Explore tab heavily rely on languages, so I think the solution to incorrect languages could be #17073.

**Comment by gdinwiddie** [comment]
I have the same problem. Post https://mstdn.jp/@neeko/109257552636605699 is one example. There is no English in the post, and Neeko's profile (https://mstdn.jp/@neeko) doesn't appear to mention English, at least not in English.

**Comment by Brawaru** [comment]
@gdinwiddie This server is using 3.5.1 which already has language detection removed. The post is published in English language: you can check that in [raw JSON](https://mstdn.jp/@neeko/109257552636605699.json) representation, through `contentMap` field. I suppose because in their settings they have posting language set to English. In newer versions (3.5.3) language dropdown was added allowing to select language per post.

**Comment by isobelws** [comment]
I'm seeing this too. Examples:
https://pl22.telteltel.com/notice/APl5eVKGAX1OjmltRo
https://mstdn.jp/@Kisaragineru/109369962060280233
https://iyasaretai.pw/@Ruhuna/109369959874521994
https://masto.nu/@amralomari/109369961726576315
https://mstdn.jp/@Kisaragineru/109369956531667912
https://mstdn.jp/@Kisaragineru/109369956033343951
https://mstdn.jp/@Xider/109369949789030511

**Comment by ronilaukkarinen** [comment]
> I'm seeing this too. Examples: https://pl22.telteltel.com/notice/APl5eVKGAX1OjmltRo https://mstdn.jp/@Kisaragineru/109369962060280233 https://iyasaretai.pw/@Ruhuna/109369959874521994 https://masto.nu/@amralomari/109369961726576315 https://mstdn.jp/@Kisaragineru/109369956531667912 https://mstdn.jp/@Kisaragineru/109369956033343951 https://mstdn.jp/@Xider/109369949789030511

It seems these are tagged `"contentMap":{"en"`: https://iyasaretai.pw/@Ruhuna/109369959874521994.json

Not everybody use the lang tag feature.

**Comment by cmeury** [comment]
Looking at the JSON response for a (federated) status, I see `"language": "es"`, but in my profile settings I only have selected: `de`, `en`, `eo`, and `fr`.

**Comment by mAcf00bar** [comment]
Still today, the language filter is not working.
Even though I have select only english, I still see toots in different languages in my feed (and yes, I've checked if those toots have set a different language than english)

**Comment by drpaulralph** [comment]
Definitely still broken, and definitely still turning off users. I have my profile set to English only and here are some excerpts from my home timeline. 

<img width="316" alt="Screenshot 2023-07-02 at 15 37 33" src="https://github.com/mastodon/mastodon/assets/33912489/69c19ccc-62c5-4fcd-bba0-763f1588309d">
<img width="314" alt="Screenshot 2023-07-02 at 15 37 52" src="https://github.com/mastodon/mastodon/assets/33912489/8f123191-df9d-4463-9c41-d477e6214e1f">
<img width="312" alt="Screenshot 2023-07-02 at 15 38 36" src="https://github.com/mastodon/mastodon/assets/33912489/42b8e94d-a23e-416a-b07c-ea236e7748dc">

**Comment by brawaru** [comment]
Mastodon has removed language detection for posts, in these scenarios the posters just didn't set their post languages correctly.

**Comment by brendanjones** [comment]
How does one check what language a post is using? This issue is definitely a problem for me but I don't know if it's a bug or people just aren't setting their post language correctly.

**Comment by brawaru** [comment]
You can check the post language:

- By simply pressing the reply button. Usually, this sets your post language to the language of the post you're replying to.

- Using the Inspect Element (<kbd>F12</kbd> / <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>I</kbd>) and finding the element for the post, then looking at the `lang` attribute.

- Copying the link to the original post and adding `.json` at the end to see the raw ActivityPub JSON representation. In the `contentMap`, you will find the post content mapped to a specific language tag. For example, in this [tagesschau](https://ard.social/@tagesschau/110714707143949037.json) post, it is marked as German, as indicated by the `"de"` key in the `"contentMap"`.

**Comment by brendanjones** [comment]
Thanks @brawaru. Righto, my language filter is working for federated timeline. I can't test it for local timeline because my instance is English-only.

The filter *doesn't* work for my home feed, I get other languages there (confirmed by looking at the json of posts). Which I guess is not a technical bug, since the setting is to filter languages in 'public timelines', and home is not a public timeline? Is that correct?

I'd call it an unexpected-behaviour bug, though. If I set a language filter I'd expect it to filter Home timeline as well.

**Comment by brendanjones** [comment]
Update: is anyone seeing posts in the home feed that don't follow their language filters, and aren't from boosts https://github.com/mastodon/mastodon/issues/20241 or followed hashtags https://github.com/mastodon/mastodon/issues/20937?

Because I've realised that all the posts I can find in my home feed that don't respect my language filters are because of those two issues. Said another way: I'm not seeing any posts from people I follow that don't respect my language filters. Only posts from boosts and followed hashtags.

**Comment by drpaulralph** [comment]
> Mastodon has removed automatic language detection for posts, in above cases the posters just didn't set their posts languages correctly.

I still have posts in my timeline where users have languages set to something other than English. Here's an example: 

<img width="309" alt="Screenshot 2023-07-16 at 14 36 19" src="https://github.com/mastodon/mastodon/assets/33912489/6c3af5e5-7ac1-4e50-8d07-9e90d9107a1f">

Yes I know this is a boost but what difference does that make? If someone says "only show me posts in languages X Y and Z" then that should also apply to boosts.

**Comment by brendanjones** [comment]
> Yes I know this is a boost but what difference does that make?

@drpaulralph Indeed, it doesn’t make a difference from a user perspective, everything in home feed should respect language filters. It is however useful to know specifically which type of posts to which to the issue applies, in order to solve it easier. 

And also to have a cleaner backlog; if the issue only applies to posts from boosts and tags then this ticket can be closed in favour of the other two I mentioned above.

**Comment by drpaulralph** [comment]
@brendanjones Fair point. All of the posts that have bypassed the language filter on my home feed in the past few days were boosts.

**Comment by KyberNull** [comment]
#20898 is also described similar issue

**Comment by kokosowy** [comment]
This issue makes using popular hashtags completely unusable on Mastodon. For example, if someone wants to follow a popular hashtag like #apple, they will receive a massive amount of posts in languages that they can't understand. This language filtering issue makes Mastodon completely useless for following such tags. I'd say it's quite a problem.

**Comment by ChasBelov** [comment]
If your browser supports a user style sheet or provides some way to add CSS through a plug-in, the following CSS will suppress posts that are marked as not being in your language. The risk is that a post will be incorrectly marked for language, but my personal philosophy is that Mastodon is a firehose and you can only read a small percentage of posts anyway.

Here's the CSS, set to hide content that is not marked as being in English. If you want to keep a different language, replace "en" with the code for your language. This code is not guaranteed to hide posts in languages that Mastodon can't translate; I'll try to update this post if I run across such a post on Mastodon.

(Updated 3/18/26)

```
#mastodon div.status:has(div.status__content__text--visible:not([lang^="en"])), 
#mastodon div.detailed-status:has(div.status__content__text--visible:not([lang^="en"])) {
    display:none !important;
}
```

**Comment by Athanasius** [comment]
> If your browser supports a user style sheet or provides some way to add CSS through a plug-in, the following CSS will suppress posts that are marked as not being in your language. The risk is that a post will be incorrectly marked for language, but my personal philosophy is that Mastodon is a firehose and you can only read a small percentage of posts anyway.
> 
> Here's the CSS, set to hide content that is not marked as being in English. If you want to keep a different language, replace "en" with the code for your language. This code is not guaranteed to hide posts in languages that Mastodon can't translate; I'll try to update this post if I run across such a post on Mastodon.
> 
> (Updated 3/18/26)
> 
> ```
> #mastodon div.status:has(div.status__content__text--visible:not([lang^="en"])), 
> #mastodon div.detailed-status:has(div.status__content__text--visible:not([lang^="en"])) {
>     display:none !important;
> }
> ```

Unfortunately this doesn't work well with toots that don't have a language set.   My specific example for this is the fedimeteo.com bot posts.   They're literally missing the `lang="..."` on `div.status__content__text--visible`.  This might technically be an API violation by them, I've not checked.

**Comment by Athanasius** [comment]
Aha!  This works:
```css
    #mastodon div.status:has(div.status__content__text--visible[lang]:not([lang^="en"])), 
#mastodon div.detailed-status:has(div.status__content__text--visible[lang]:not([lang^="en"])) {
        display:none !important;
    }
```
